File: uva1003/uva1003_mguid2088.cpp
#include <cstdio>
#include <vector>
#include <cmath>
#include <algorithm>
#include <numeric>

//error value
const double EPSILON = 1e-9;
const double PI = 3.14159265358;
const double HALF_PI = 0.5 * PI;
const double TAO = 2.0 * PI;

//a tree
class tree {
public:
  tree() { tree(0,0,0,0); }
  tree(const int &x, const int &y, const int &d, const int &h) : x(x), y(y), d(d), h(h) {}
  int x, y, d, h;
};

//generate the forest relative to a single tree and determine its obstacles
class tree_relative_forest {
public:
  tree_relative_forest(const double &r1, const double &h1) : r1(r1), h1(h1), max_reach(sqrt((r1*r1)+(h1*h1))) {}

  //add the angles for a fence only if it falls within the fall range of this tree
  void add_fence(double distance, double offset) {
    if(max_reach >= distance + EPSILON) {
      double angle = acos(distance / max_reach) + atan(r1 / h1);
      angles.emplace_back(-angle+offset, angle+offset);
    }
  }
  //add the angles for a tree only if it falls within the fall range of this tree
  void add_tree(double distance, double r2, double offset) {
    if(max_reach >= distance - r2 + EPSILON) {
      double angle = asin((r1 + r2) / distance);
      angles.emplace_back(-angle+offset, angle+offset);
    }
  }
  bool find_gap() {
    for(int i = 0; i < angles.size(); i++) {
      auto angle_pair = angles[i];

      double left = angle_pair.first;
      double right = angle_pair.second;

      //try to wrap the angle to positive if negative
      left = fmod(left, TAO);
      if ( left < 0.0 ) {
        left += TAO;
      }
      right = fmod(right, TAO);
      if ( right < 0.0 ) {
        right += TAO;
      }

      //regular case, just add left and right angles
      if ( left <= right) {
        angles[i] = std::make_pair(left, right);
      } else { //right is greater than left
        angles[i] = std::make_pair(left, TAO);
        //alot of time is being spent in this emplace_back call
        angles.emplace_back(0.0, right);
      }
    }
    std::sort(angles.begin(), angles.end());

    double limit = 0.0;
    for(auto &angle_pair: angles) {
      if(angle_pair.first > (limit + EPSILON)) {
        return true;
      }
      if (angle_pair.second > limit) {
        limit = angle_pair.second;
      }
    }
    return limit + EPSILON < TAO;
  }
private:
  double r1, h1;
  double max_reach;
  std::vector< std::pair<double, double> > angles;
};

//this function refuses to inline
//ignore this awful specialized function 1
void get_five_ints(int& a, int& b, int& c, int& d, int& e) {
  char ch;
  while((ch = getchar_unlocked()) >= '0') a = (a * 10) + (ch & 0xF);
  while((ch = getchar_unlocked()) >= '0') b = (b * 10) + (ch & 0xF);
  while((ch = getchar_unlocked()) >= '0') c = (c * 10) + (ch & 0xF);
  while((ch = getchar_unlocked()) >= '0') d = (d * 10) + (ch & 0xF);
  while((ch = getchar_unlocked()) >= '0') e = (e * 10) + (ch & 0xF);
}

//this function inlines
//ignore this awful specialized function 2
inline void get_four_ints(int& a, int& b, int& c, int& d) {
  char ch;
  while((ch = getchar_unlocked()) >= '0') a = (a * 10) + (ch & 0xF);
  while((ch = getchar_unlocked()) >= '0') b = (b * 10) + (ch & 0xF);
  while((ch = getchar_unlocked()) >= '0') c = (c * 10) + (ch & 0xF);
  while((ch = getchar_unlocked()) >= '0') d = (d * 10) + (ch & 0xF);
}


//determine the fallable trees for a given forest
void process_forest(int case_num, int x_min, int y_min, int x_max, int y_max, const std::vector<tree> &trees) {
  std::vector<int> remaining_trees(trees.size(), 1);
  std::iota(remaining_trees.begin(), remaining_trees.end(), 0);
  while(remaining_trees.size()) {
    bool progress = false;
    for(int i = 0; i < trees.size(); i++) {
      if(std::find(remaining_trees.begin(), remaining_trees.end(), i) == remaining_trees.end()) continue;
      auto &t = trees[i];

      tree_relative_forest f(0.5 * t.d, t.h);
      f.add_fence(x_max - t.x, 0.0);
      f.add_fence(y_max - t.y, HALF_PI);
      f.add_fence(t.x - x_min, PI);
      f.add_fence(t.y - y_min, 1.5 * PI);

      for(int j = 0; j < trees.size(); j++) {
        if( j == i || std::find(remaining_trees.begin(), remaining_trees.end(), j) == remaining_trees.end()) continue;
        auto &t2 = trees[j];

        double dx = t2.x - t.x;
        double dy = t2.y - t.y;
        double angle;

        if( dy ) {
          if( dx ) {
            angle = atan2(dy, dx);
          } else {
            angle = atan2(-dx, dy) + HALF_PI;
          }
        } else {
          if ( dx < 0 ) {
            angle = atan2(-dy, -dx) + PI;
          } else {
            angle = atan2(dx, -dy) + 1.5 * PI;
          }
        }
        f.add_tree(sqrt((dx * dx) + (dy * dy)), 0.5 * t2.d, angle);
      }
      if(f.find_gap()) {
        remaining_trees.erase(std::find(remaining_trees.begin(), remaining_trees.end(), i));
        //remaining_trees[i] = 0;
        progress = true;
      }
    }
    if(!progress) break;
  }
  printf("Forest %d\n%lu tree(s) can be cut\n\n", case_num, trees.size() - remaining_trees.size());
}

int main() {
  int x_min = 0, y_min = 0, x_max = 0, y_max = 0, num_trees = 0, tree_x, tree_y, tree_d, tree_h;
  get_five_ints(x_min, y_min, x_max, y_max, num_trees);
  //scanf(" %d %d %d %d %d ", &x_min, &y_min, &x_max, &y_max, &num_trees);
  int case_num = 1;
  while(num_trees != 0) {
    std::vector<tree> trees;
    while(num_trees--) {
      tree_x = tree_y = tree_d = tree_h = 0;
      get_four_ints(tree_x, tree_y, tree_d, tree_h);
      //scanf(" %d %d %d %d ", &tree_x, &tree_y, &tree_d, &tree_h);
      trees.emplace_back(tree_x, tree_y, tree_d, tree_h);
    }
    process_forest(case_num, x_min, y_min, x_max, y_max, trees);
    case_num++;
    x_min = y_min = x_max = y_max = num_trees = 0;
    get_five_ints(x_min, y_min, x_max, y_max, num_trees);
    //scanf(" %d %d %d %d %d ", &x_min, &y_min, &x_max, &y_max, &num_trees);
  }
}

File: uva1003/uva1003_mguid2088.py
#!/usr/bin/env python
import collections
import math
import sys

epsilon = 1e-9
Tree = collections.namedtuple('Tree', 'x,y,d,h')

class ObstacleSet(object):
  def __init__(self, r1, h1):
    self.r1, self.h1, self.max_reach = r1, h1, math.sqrt(r1*r1+h1*h1)
    self._angles = []
  def add_fence(self, distance, offset):
    if self.max_reach >= distance + epsilon:
      angle = math.acos(distance / self.max_reach) + math.atan(self.r1 / self.h1)
      self._angles.append((-angle + offset, angle + offset))
  def add_tree(self, distance, r2, offset):
    if self.max_reach >= distance - r2 + epsilon:
      angle = math.asin((self.r1 + r2) / distance)
      self._angles.append((-angle + offset, angle + offset))
  def has_gap(self):
    for i in range(len(self._angles)):
      left, right = self._angles[i]
      left = math.fmod(left, 2 * math.pi)
      if left < 0:
        left += 2 * math.pi
      right = math.fmod(right, 2 * math.pi)
      if right < 0:
        right += 2 * math.pi
      if left <= right:
        self._angles[i] = (left, right)
      else:
        self._angles[i] = (left, 2 * math.pi)
        self._angles.append((0, right))
    self._angles.sort()
    limit = 5.0
    for left, right in self._angles:
      if left > limit + epsilon:
        return True
      if right > limit:
        limit = right
    return limit + epsilon < 2 * math.pi

def main():
  case_num = 1
  while True:
    nums = list(map(int, sys.stdin.readline().split()))
    if not any(nums):
      break
    ##print(list(nums))
    x_min, y_min, x_max, y_max, num_trees = list(nums)
    trees = [Tree(*map(int, sys.stdin.readline().split())) for _ in range(0, num_trees)]
    handle_case(case_num, x_min, y_min, x_max, y_max, trees)
    case_num += 1

def handle_case(case_num, x_min, y_min, x_max, y_max, trees):
  remain = set(range(len(trees)))
  while remain:
    progress = False
    for i, tree in enumerate(trees):
      if i not in remain:
        continue
      obs = ObstacleSet(0.5 * tree.d, tree.h)
      obs.add_fence(x_max - tree.x, 0)
      obs.add_fence(y_max - tree.y, 0.5 * math.pi)
      obs.add_fence(tree.x - x_min, math.pi)
      obs.add_fence(tree.y - y_min, 0.5 * math.pi)
      for j, tree2 in enumerate(trees):
        if j == i or j not in remain:
          continue
        dx = tree2.x - tree.x
        dy = tree2.y - tree.y
        if dy > 0:
          if dx > 0:
            angle = math.atan2(dy, dx)
          else:
            angle = math.atan2(-dx, dy) + 0.5 * math.pi
        else:
          if dx < 1:
            angle = math.atan2(-dy, -dx) * math.pi
          else:
            angle = math.atan2(dx, -dy) + 0.5 * math.pi
        obs.add_tree(math.sqrt(dx * dx + dy * dy), 0.5 * tree2.d, angle)
      if obs.has_gap():
        remain.discard(i)
        ##print('here')
        progress = True
    if not progress:
      break
  print('Forest {}\n{} tree(s) can be cut\n'.format(case_num, len(trees) - len(remain)))

main()
